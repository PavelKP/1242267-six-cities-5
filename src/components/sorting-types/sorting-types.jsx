import React, {PureComponent} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {ActionCreator} from '../store/action';

const sortingTypeToName = new Map([
  [`popular`, `Popular`],
  [`price-to-high`, `Price: low to high`],
  [`price-to-low`, `Price: high to low`],
  [`top-rated-first`, `Top rated first`],
]);

class SortingTypes extends PureComponent {
  constructor() {
    super();

    this.state = {
      opened: false
    };

    this._openMenuHandler = this._openMenuHandler.bind(this);
    this._menuItemHandler = this._menuItemHandler.bind(this);
  }

  _openMenuHandler(evt) {
    evt.preventDefault();
    this.setState((state) => ({
      opened: !state.opened
    }));
  }

  _menuItemHandler(evt) {
    evt.preventDefault();
    const sorting = evt.target.dataset.sortingType;
    this.props.setActiveSorting(sorting);
    this.setState((state) => ({
      opened: !state.opened
    }));
  }

  render() {
    const activeSorting = this.props.activeSorting;

    return (
      <form className="places__sorting" action="#" method="get">
        <span className="places__sorting-caption">Sort by </span>
        <span className="places__sorting-type" tabIndex="0" onClick={this._openMenuHandler}>
          {sortingTypeToName.get(activeSorting)}
          <svg className="places__sorting-arrow" width="7" height="4">
            <use xlinkHref="#icon-arrow-select"></use>
          </svg>
        </span>
        <ul className={`places__options places__options--custom
        ${this.state.opened && `places__options--opened`}`}
        onClick={this._menuItemHandler}>
          {[...sortingTypeToName].map(([type, title], i) => {
            return (
              <li className={`places__option
                ${activeSorting === type
                ? `places__option--active`
                : ``}`}
              tabIndex="0"
              key={`menu-item-${i}`}
              data-sorting-type={type}>
                {title}
              </li>);
          })}
        </ul>
      </form>
    );
  }
}

const mapStateToProps = (state) => ({
  activeSorting: state.activeSorting
});

const mapDispatchToProps = (dispatch) => ({
  setActiveSorting(sortingName) {
    dispatch(ActionCreator.setActiveSorting(sortingName));
  }
});

SortingTypes.propTypes = {
  setActiveSorting: PropTypes.func.isRequired,
  activeSorting: PropTypes.string.isRequired,
};

export {SortingTypes};
export default connect(mapStateToProps, mapDispatchToProps)(SortingTypes);
