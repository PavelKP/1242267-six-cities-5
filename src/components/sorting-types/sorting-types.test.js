import React from 'react';
import renderer from 'react-test-renderer';
import {SortingTypes} from './sorting-types';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import mockStoreData from '../../__mocks__/mock-store';
import {Router} from 'react-router-dom';
import browserHistory from '../../browser-history';

const mockStore = configureStore([]);

describe(`<SortingTypes /> render`, () => {

  let store;

  beforeEach(() => {
    store = mockStore(mockStoreData);
  });

  it(`Should SortingTypes render correctly`, ()=> {
    const tree = renderer.create(
        <Provider store={store}>
          <Router history={browserHistory}>
            <SortingTypes
              setActiveSorting={()=>{}}
              activeSorting={`popular`}
              togglerState={false}
              onTogglerClick={()=>{}}
            />
          </Router>
        </Provider>)
        .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should SortingTypes with opened menu render correctly`, ()=> {
    const tree = renderer.create(
        <Provider store={store}>
          <Router history={browserHistory}>
            <SortingTypes
              setActiveSorting={()=>{}}
              activeSorting={`popular`}
              togglerState={true}
              onTogglerClick={()=>{}}
            />
          </Router>
        </Provider>)
        .toJSON();

    expect(tree).toMatchSnapshot();
  });

});
