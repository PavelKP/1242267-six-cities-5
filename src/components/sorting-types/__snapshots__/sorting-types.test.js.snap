// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SortingTypes /> render Should SortingTypes render correctly 1`] = `
<form
  action="#"
  className="places__sorting"
  method="get"
>
  <span
    className="places__sorting-caption"
  >
    Sort by 
  </span>
  <span
    className="places__sorting-type"
    onClick={[Function]}
    tabIndex="0"
  >
    Popular
    <svg
      className="places__sorting-arrow"
      height="4"
      width="7"
    >
      <use
        xlinkHref="#icon-arrow-select"
      />
    </svg>
  </span>
  <ul
    className="places__options places__options--custom
        "
    onClick={[Function]}
  >
    <li
      className="places__option
              places__option--active"
      data-sorting-type="popular"
      tabIndex="0"
    >
      Popular
    </li>
    <li
      className="places__option
              "
      data-sorting-type="price-to-high"
      tabIndex="0"
    >
      Price: low to high
    </li>
    <li
      className="places__option
              "
      data-sorting-type="price-to-low"
      tabIndex="0"
    >
      Price: high to low
    </li>
    <li
      className="places__option
              "
      data-sorting-type="top-rated-first"
      tabIndex="0"
    >
      Top rated first
    </li>
  </ul>
</form>
`;

exports[`<SortingTypes /> render Should SortingTypes with opened menu render correctly 1`] = `
<form
  action="#"
  className="places__sorting"
  method="get"
>
  <span
    className="places__sorting-caption"
  >
    Sort by 
  </span>
  <span
    className="places__sorting-type"
    onClick={[Function]}
    tabIndex="0"
  >
    Popular
    <svg
      className="places__sorting-arrow"
      height="4"
      width="7"
    >
      <use
        xlinkHref="#icon-arrow-select"
      />
    </svg>
  </span>
  <ul
    className="places__options places__options--custom
        places__options--opened"
    onClick={[Function]}
  >
    <li
      className="places__option
              places__option--active"
      data-sorting-type="popular"
      tabIndex="0"
    >
      Popular
    </li>
    <li
      className="places__option
              "
      data-sorting-type="price-to-high"
      tabIndex="0"
    >
      Price: low to high
    </li>
    <li
      className="places__option
              "
      data-sorting-type="price-to-low"
      tabIndex="0"
    >
      Price: high to low
    </li>
    <li
      className="places__option
              "
      data-sorting-type="top-rated-first"
      tabIndex="0"
    >
      Top rated first
    </li>
  </ul>
</form>
`;
